File: 0383-ransom-note/0383-ransom-note.js
function canConstruct(ransomNote, magazine) {
    if (ransomNote.length > magazine.length) return false;
    const counts = new Int32Array(26);

    for (let i = 0; i < magazine.length; i++) {
        counts[magazine.charCodeAt(i) - 97]++;
    }

    for (let i = 0; i < ransomNote.length; i++) {
        const index = ransomNote.charCodeAt(i) - 97;
        counts[index]--;
        if (counts[index] < 0) return false;
    }
    return true;
}

File: 0383-ransom-note/0383-ransom-note.py
from collections import Counter

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(ransomNote) > len(magazine):
            return False 
        
        mag_counts = Counter(magazine)
        
        for char in ransomNote:
            if mag_counts[char] <= 0:
                return False
            mag_counts[char] -= 1
            
        return True
